R/grid.R

```diff
@@ -0,0 +1,295 @@
+#'    {custom_style_grid_container}"
+#' }
+#'
+grid_container_css <- function(css_grid_template_areas, rows_height, cols_width) {
+  grid_container_styles <- c(
+    "display: grid",
+    "height: 100%",
+    glue::glue("grid-template-rows: {paste(rows_height, collapse = ' ')}"),
+    glue::glue("grid-template-columns: {paste(cols_width, collapse = ' ')}"),
+    glue::glue("grid-template-areas: {css_grid_template_areas}"),
+    "{custom_style_grid_container}" # To be rendered by glue::glue() later
+  )
+  paste(grid_container_styles, collapse = "; ")
+}
+
+#' Generate list of HTML div elements representing grid areas.
+#'
+#' @param area_names
+#'
+#' @return list of shiny::tags$div
+#'
+#' @details This is a helper function used in grid_template()
+#' \preformatted{
+#'   list_of_area_tags(c("header", "main", "footer"))
+#' }
+#' returns the following list:
+#' \preformatted{
+#'   [[1]] <div style="grid-area: header; {custom_style_grid_area_header}">{{ header }}</div>
+#'   [[2]] <div style="grid-area: main; {custom_style_grid_area_main}">{{ main }}</div>
+#'   [[3]] <div style="grid-area: footer; {custom_style_grid_area_footer}">{{ footer }}</div>
+#' }
+#'
+list_of_area_tags <- function(area_names) {
+  lapply(area_names,
+    function(name) shiny::tags$div(
+     style = as.character(glue::glue("grid-area: {name}; {{custom_style_grid_area_{name}}}")), paste("{{", name, "}}")
+    )
+  )
+}
+
+#' Define a template of a CSS grid
+#'
+#' @param default
+#' Template for desktop:
+#' list(areas = [data.frame of strings],
+#'      rows_height = [vector of strings],
+#'      cols_width = [vector of strings])
+#' @param mobile
+#' Template for mobile:
+#' list(areas = [data.frame of strings],
+#'      rows_height = [vector of strings],
+#'      cols_width = [vector of strings])
+#'
+#' @return
+#' list(template = [string], area_names = [vector of strings])
+#'
+#' template - contains template that can be formatted with glue::glue() function
+#'
+#' area_names - contain all unique area names used in grid definition
+#'
+#' @details
+#' Example usage:
+#' \preformatted{
+#'   myGrid <- grid_template(default = list(
+#'     areas = rbind(
+#'       c("header", "header", "header"),
+#'       c("menu",   "main",   "right1"),
+#'       c("menu",   "main",   "right2")
+#'     ),
+#'     rows_height = c("50px", "auto", "100px"),
+#'     cols_width = c("100px", "2fr", "1fr")
+#'   ))
+#'
+#'   display_grid(myGrid)
+#' }
+#' Second example:
+#' \preformatted{
+#'   subGrid <- grid_template(default = list(
+#'     areas = rbind(
+#'       c("top_left", "top_right"),
+#'       c("bottom_left", "bottom_right")
+#'     ),
+#'     rows_height = c("50\%", "50\%"),
+#'     cols_width = c("50\%", "50\%")
+#'   ))
+#'
+#'   display_grid(subGrid)
+#' }
+#'
+#' @export
+#'
+grid_template <- function(
+  default = list(areas = "", rows_height = c(), cols_width = c()),
+  mobile = list(areas = "", rows_height = c(), cols_width = c())) {
```

Comment:
Here I was thinking: how to make it most flexible, but at the same time user-friendly. We talked about it briefly and you mentioned yourself that there's a handful of other scenarios to consider, like `tablet` format for example... Maybe unnamed parameters would do (`...`), assuming that they have specified format.

Also, is there a plan to add in `grid_template` the conditions for switching between mobile and default?

Reply:
Use of ellipsis (`...`) here should be considered after we implement the logic that is responsible for switching between grids depending on the screen size. This is a good idea, but let's see how our logic works first.

In the simplest implementation this logic for now can decide when it displays mobile (hardcoded rules for screen sizes, "if width <800px then mobile"). Later we can allow user to define their own rules / media queries. But it has to be easy to learn and use for R users. I see at least two ideas:

* it can be an additional parameter on the `list(areas = ...)` object e.g. `max.width` and `min.width` 
* it can be argument e.g. `media.rules = list(mobile = "800px", tablet = "1200px")`

The logic for switching between mobile and default should use the following technique of combining media queries `@media` and css grid: https://www.w3.org/WAI/WCAG21/Techniques/css/C32.html